lib/sqlitex.ex

```diff
@@ -7,19 +7,47 @@ defmodule Sqlitex do
     :esqlite3.open(path)
   end
 
-  def query(db, query, opts \\ []) do
-    {:ok, statement} = :esqlite3.prepare(query, db)
-    types = :esqlite3.column_types(statement) |> Tuple.to_list
-    columns = :esqlite3.column_names(statement) |> Tuple.to_list
-    rows = :esqlite3.fetchall(statement)
-    into = Keyword.get(opts, :into, [])
-    Sqlitex.Row.from(types, columns, rows, into)
-  end
-
   def with_db(path, fun) do
     {:ok, db} = open(path)
     res = fun.(db)
     close(db)
     res
   end
+
+  def exec(db, sql) do
+    :esqlite3.exec(sql, db)
+  end
+
+  def query(db, sql, opts \\ []) do
+    {params, into} = query_options(opts)
+    {:ok, statement} = :esqlite3.prepare(sql, db)
+    :ok = :esqlite3.bind(statement, params)
+    types = :esqlite3.column_types(statement)
+    columns = :esqlite3.column_names(statement)
+    rows = :esqlite3.fetchall(statement)
+    return_rows_or_error(types, columns, rows, into)
+  end
+
+  defp query_options(opts) do
+    params = Keyword.get(opts, :bind, [])
+    into = Keyword.get(opts, :into, [])
+    {params, into}
+  end
+
+  defp return_rows_or_error({:error, kind} = error, columns, rows, into) do
+    case kind do
+      :no_columns -> return_rows_or_error({}, columns, rows, into)
+      _ -> error
+    end
+  end
+  defp return_rows_or_error(types, {:error, kind} = error, rows, into) do
+    case kind do
+      :no_columns -> return_rows_or_error(types, {}, rows, into)
+      _ -> error
+    end
+  end
```

Comment:
I think you're right [it looks like esqlite fabricates this error](https://github.com/mmzeeman/esqlite/blob/58e5e6ed55f45d3813bfc4bd2e966178809b390e/c_src/esqlite3_nif.c#L550) rather than passing it along from the sqlite3 source code